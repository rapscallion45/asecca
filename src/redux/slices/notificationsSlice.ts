import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import {
  INotificationState,
  IAddNotificationPayload,
  ICloseNotificationPayload,
  IRemoveNotificationPayload,
} from '../types';

/**
 * Global application state slice definition for Notifications
 *
 * Notifications are the alert popups seen by the user on specific action
 * results, such as errors with API requests, login failures, etc. This state
 * is manipulated from multiple places in the application.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 */

/**
 * Initial Notifications State
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IInitialNotificationsState
 * @prop {Array<INotificationState>} data - notification state data
 */
interface IInitialNotificationsState {
  data: Array<INotificationState>;
}

/**
 * Initialise notification state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @constant
 * @type {IInitialNotificationsState}
 */
const initialNotificationsState: IInitialNotificationsState = {
  data: [],
};

/**
 * Create the redux slice for interacting with the notifications state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const notificationsSlice = createSlice({
  name: 'notifications',
  initialState: initialNotificationsState,
  reducers: {
    addNotification: (
      state,
      action: PayloadAction<IAddNotificationPayload>
    ) => ({
      ...state,
      data: [
        ...state.data,
        {
          message: action.payload.message,
          dismissed: false,
          options: {
            key: new Date().getTime() + Math.random(),
            variant: action.payload.variant,
          },
        },
      ],
    }),
    closeNotification: (
      state,
      action: PayloadAction<ICloseNotificationPayload>
    ) => ({
      ...state,
      data: state.data.map((notification: INotificationState) =>
        notification.options.key === action.payload.key
          ? { ...notification, dismissed: true }
          : { ...notification }
      ),
    }),
    removeNotification: (
      state,
      action: PayloadAction<IRemoveNotificationPayload>
    ) => ({
      ...state,
      data: state.data.filter(
        (notification: INotificationState) =>
          notification.options.key !== action.payload.key
      ),
    }),
  },
});

/**
 * Notification actions for adding, closing and removing a notifications
 *
 * @since 0.0.0
 */
export const { addNotification, closeNotification, removeNotification } =
  notificationsSlice.actions;

export default notificationsSlice.reducer;
