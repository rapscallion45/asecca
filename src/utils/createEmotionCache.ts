import createCache from '@emotion/cache';

/* create emotion style cache, and set 'prepend' true to allow MUI style overrides */

const createEmotionCache = () => createCache({ key: 'css', prepend: true });

export default createEmotionCache;
