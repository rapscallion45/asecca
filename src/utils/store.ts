/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
