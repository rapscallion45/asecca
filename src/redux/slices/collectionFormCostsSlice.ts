import {
  ActionReducerMapBuilder,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import {
  ICollectionFormCostsData,
  ICollectionFormCostsDataPayload,
} from '@/lib/api/api-types';
import collectionFormService from '../../services/forms/collectionFormService';
import {
  ICollectionFormCostsState,
  IFetchCollectionFormCostsByCollectionIdArgs,
  ISaveCollectionFormCostsByCollectionIdArgs,
  ICollectionFormCostsEditCostsPayload,
} from '../types';
import { addNotification } from './notificationsSlice';

/**
 * State slice definition for Collection Form Costs
 *
 * All state updates to the Collection Form Costs are managed here.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 */

/**
 * Async thunk for GET /api/collection/costs/api/costs API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchByCollectionId = createAsyncThunk(
  'collectionFormCosts/fetchByCollectionId',
  async (args: IFetchCollectionFormCostsByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getCosts(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Collection Form Costs from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for POST /api/collection/costs/api/costs API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const saveByCollectionId = createAsyncThunk(
  'collectionFormCosts/saveByCollectionId',
  async (args: ISaveCollectionFormCostsByCollectionIdArgs, thunkAPI) => {
    const res = await collectionFormService.setCosts(args.data);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to save Collection Form Costs to server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    } else {
      thunkAPI.dispatch(
        addNotification({
          message: 'Successfully saved Collection Form Costs to server',
          variant: 'success',
        })
      );
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Initialises Collection Form Costs state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @constant
 * @type {ICollectionFormCostsState}
 */
const initialCollectionFormCostsState: ICollectionFormCostsState = {
  loading: false,
  data: { rows: [] },
  dataShadow: { rows: [] },
  saving: false,
  edited: false,
};

/**
 * Create the redux slice for interacting with the Collection Form Costs state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const collectionFormCostsSlice = createSlice({
  name: 'collectionFormCosts',
  initialState: initialCollectionFormCostsState,
  reducers: {
    /* reducer used for user input changes to the Costs data */
    editCosts: (
      state,
      action: PayloadAction<ICollectionFormCostsEditCostsPayload>
    ) => {
      /* find and update passed cost */
      state.data.rows = state.data.rows.map(
        (cost: ICollectionFormCostsData, index: number) => {
          /* perform update for passed table row number */
          if (index === action.payload.rowIdx) {
            return {
              ...cost,
              /* update the value of the passed column */
              [action.payload.colKey as keyof ICollectionFormCostsData]:
                action.payload.value,
            };
          }
          return cost;
        }
      );
      state.edited = true;
    },
    /* reducer used for when user clears edits to Costs data */
    resetCosts: (state) => {
      /* reset the data by simply copying the shadow to working copy */
      state.data = state.dataShadow;
      state.edited = false;
    },
  },
  extraReducers: (
    builder: ActionReducerMapBuilder<ICollectionFormCostsState>
  ) => {
    /*
     * `extraReducers` used to handle actions that were generated
     * outside of the Collection Form costs slice, such as the async thunks
     * for the API requests, defined at the top of this file.
     */
    builder
      /* Fetch Collection Form Costs extra reducers */
      .addCase(
        fetchByCollectionId.pending,
        (state: ICollectionFormCostsState) => {
          state.loading = true;
          state.data = { rows: [] };
          state.dataShadow = { rows: [] };
          state.error = undefined;
          state.edited = false;
        }
      )
      .addCase(
        fetchByCollectionId.fulfilled,
        (
          state: ICollectionFormCostsState,
          action: PayloadAction<ICollectionFormCostsDataPayload>
        ) => {
          state.loading = false;
          state.data = action.payload;
          state.dataShadow = action.payload;
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.rejected,
        (state: ICollectionFormCostsState) => {
          state.loading = false;
          state.data = { rows: [] };
          state.dataShadow = { rows: [] };
          state.error =
            'Failed to load Collection Form Costs data from server.';
        }
      )
      /* Save Collection Form Costs extra reducers */
      .addCase(
        saveByCollectionId.pending,
        (state: ICollectionFormCostsState) => {
          state.saving = true;
        }
      )
      .addCase(
        saveByCollectionId.fulfilled,
        (state: ICollectionFormCostsState) => {
          state.saving = false;
          state.edited = false;
        }
      )
      .addCase(
        saveByCollectionId.rejected,
        (state: ICollectionFormCostsState) => {
          state.saving = false;
        }
      );
  },
});

/* Collection Form Costs actions for editing and resetting costs data */
export const { editCosts, resetCosts } = collectionFormCostsSlice.actions;

export default collectionFormCostsSlice.reducer;
