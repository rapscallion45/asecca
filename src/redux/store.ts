import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
