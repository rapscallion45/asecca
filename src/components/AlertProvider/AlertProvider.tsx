import { FC, ReactNode, forwardRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SnackbarProvider, SnackbarKey } from 'notistack';
import { Alert } from '@mui/material';
import { closeNotification } from '../../redux/slices/notificationsSlice';
import { AppDispatch, AppState } from '../../redux/store';
import { INotificationState } from '../../redux/types';

/**
 * Alert Notification Props
 *
 * @since - 0.0.0
 *
 * @typedef IAlertNotificationProps
 * @prop {SnackbarKey} id - ID of the passed alert notification
 */
interface IAlertNotificationProps {
  id: SnackbarKey;
}

/**
 * Alert Notification Ref type
 *
 * @since - 0.0.0
 *
 * @typedef {HTMLDivElement} Ref - ref to Alert Provider parent div
 */
type Ref = HTMLDivElement;

/**
 * Alert Notification
 *
 * @author - Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since - 0.0.0
 *
 * @param {IAlertNotificationProps} props - component props
 * @param {Ref} ref - reference to the alert provider parent component
 * @returns {ForwardRefExoticComponent} - alert notification component
 */
const AlertNotification = forwardRef<Ref, IAlertNotificationProps>(
  (props, ref) => {
    const dispatch = useDispatch<AppDispatch>();
    const { id } = props;

    /** close alert callback */
    const handleClose = () => {
      dispatch(closeNotification({ key: id }));
    };

    /** find this alert in state and get variant */
    const { data: notifications } = useSelector(
      (state: AppState) => state.notifications
    );
    const successNotification = notifications.find(
      (notification: INotificationState) => notification.options.key === id
    );

    return (
      <Alert
        ref={ref}
        elevation={6}
        variant="filled"
        severity={successNotification?.options.variant}
        onClose={handleClose}
      >
        {successNotification?.message}
      </Alert>
    );
  }
);

/**
 * Alert Provider Props
 *
 * @since - 0.0.0
 *
 * @typedef IAlertProviderProps
 * @prop {ReactNode} children - component children nodes
 */
interface IAlertProviderProps {
  children?: ReactNode;
}

/**
 * Alert Provider
 *
 * @author - Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since - 0.0.0
 *
 * @param {IAlertProviderProps} props - component props
 * @returns {FC} - alert provider functional component
 */
const AlertProvider: FC<IAlertProviderProps> = (props) => {
  const { children = null } = props;

  return (
    <SnackbarProvider
      maxSnack={5}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'center',
      }}
      /** override all alert variants with our custom component */
      Components={{
        success: AlertNotification,
        error: AlertNotification,
        info: AlertNotification,
        warning: AlertNotification,
      }}
    >
      {children}
    </SnackbarProvider>
  );
};

export default AlertProvider;
