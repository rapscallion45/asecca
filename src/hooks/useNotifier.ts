import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useSnackbar, SnackbarKey } from 'notistack';
import { AppDispatch, AppState } from '@/redux/store';
import { removeNotification } from '../redux/slices/notificationsSlice';

/**
 * local copy of currently displayed notification ID keys
 *
 * @since - 0.0.0
 */
let displayed: any = [];

/**
 * Notifier hook for processing currently active notifications held in
 * global app state.
 *
 * @author - [Carl Scrivener](https://github.com/rapscallion45)
 * @since - 0.0.0
 */
const useNotifier = () => {
  /** app dispatch shorthand helper */
  const dispatch = useDispatch<AppDispatch>();

  /** global app notification state data */
  const { data: notifications } = useSelector(
    (state: AppState) => state.notifications
  );

  /** snackbar hooks */
  const { enqueueSnackbar, closeSnackbar } = useSnackbar();

  /**
   * Store displayed notification ID key helper
   *
   * @author - [Carl Scrivener](https://github.com/rapscallion45)
   * @since - 0.0.0
   */
  const storeDisplayed = (id: SnackbarKey | undefined) => {
    displayed = [...displayed, id];
  };

  /**
   * Remove displayed notification ID key helper
   *
   * @author - [Carl Scrivener](https://github.com/rapscallion45)
   * @since - 0.0.0
   */
  const removeDisplayed = (id: SnackbarKey) => {
    displayed = [...displayed.filter((key: SnackbarKey) => id !== key)];
  };

  /** whenever notifications state changes, update displayed/removed list */
  useEffect(() => {
    notifications.forEach(({ message, options = {}, dismissed = false }) => {
      if (dismissed) {
        /** dismiss snackbar using notistack */
        closeSnackbar(options.key);
        return;
      }

      /** do nothing if snackbar is already displayed */
      if (displayed.includes(options.key)) return;

      /** display snackbar using notistack */
      enqueueSnackbar(message, {
        ...options,
        onClose: (event, reason, myKey) => {
          // @ts-ignore
          if (options.onClose) options.onClose(event, reason, myKey);
        },
        onExited: (event, myKey) => {
          /** remove this snackbar from redux store */
          dispatch(removeNotification({ key: myKey }));
          removeDisplayed(myKey);
        },
      });

      /** keep track of snackbars that we've displayed */
      storeDisplayed(options.key);
    });
  }, [notifications, closeSnackbar, enqueueSnackbar, dispatch]);
};

export default useNotifier;
