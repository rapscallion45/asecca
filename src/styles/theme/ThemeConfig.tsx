import { FC, useMemo, ReactNode } from 'react';
import { useSelector } from 'react-redux';
import { CacheProvider, EmotionCache } from '@emotion/react';
import { CssBaseline } from '@mui/material';
import {
  ThemeProvider,
  createTheme,
  Theme,
  ThemeOptions,
} from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';
import { AppState } from '@/redux/store';
import lightThemePalette from './lightTheme/lightThemePalette';
import darkThemePalette from './darkTheme/darkThemePalette';
import typography from './typography';

/**
 * Theme Config Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on creating emotion styling engine cache](https://emotion.sh/docs/introduction)
 *
 * @typedef IThemeConfigProps
 * @prop {EmotionCache} emotionCache - emotion styling engine cache
 * @prop {ReactNode} children - component child node
 */
interface IThemeConfigProps {
  emotionCache: EmotionCache;
  children?: ReactNode;
}

/**
 * Theme Config
 *
 * Global application theme definitions and setup
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @component
 * @param {IThemeConfigProps} props - component props
 * @returns {FC} - theme config functional component
 */
const ThemeConfig: FC<IThemeConfigProps> = (props) => {
  const { emotionCache, children } = props;

  /* grab global theme state */
  const { type: themeType } = useSelector((state: AppState) => state.theme);

  /* check if user has dark mode preference */
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');

  /* get options and create theme for current state */
  const themeOptions: ThemeOptions = useMemo(
    () => ({
      palette:
        themeType === 'light' && !prefersDarkMode
          ? lightThemePalette
          : darkThemePalette,
      typography,
    }),
    [themeType, prefersDarkMode]
  );
  const theme: Theme = createTheme(themeOptions);

  return (
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
};

export default ThemeConfig;
