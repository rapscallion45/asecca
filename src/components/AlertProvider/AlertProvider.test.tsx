import React, { FC } from 'react';
import { Provider } from 'react-redux';
import {
  waitFor,
  render,
  fireEvent,
  screen,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import createEmotionCache from '../../utils/createEmotionCache';
import ThemeConfig from '../../styles/theme/ThemeConfig';
import palette from '../../styles/theme/lightTheme/lightThemePalette';
import store from '../../redux/store';
import useNotifier from '../../hooks/useNotifier';
import AlertProvider from './AlertProvider';
import { addNotification } from '../../redux/slices/notificationsSlice';

/**
 * test text
 *
 * @since 0.0.0
 */
const testMessage = 'Test render alert message';

/**
 * test child component
 *
 * @since 0.0.0
 */
const TestChild: FC = () => {
  useNotifier();
  return <div>Test child</div>;
};

/**
 * Alert Provider Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 */
describe('Alert Provider', () => {
  describe('Message Text', () => {
    it('Should render the passed message', async () => {
      /** Arrange */
      store.dispatch(
        addNotification({
          message: testMessage,
          variant: 'success',
        })
      );

      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider>
              <TestChild />
            </AlertProvider>
          </ThemeConfig>
        </Provider>
      );

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByRole('alert')).toBeInTheDocument();
      });
      expect(screen.getByText(testMessage)).toBeInTheDocument();
    });
  });

  describe('Children', () => {
    it('Should render passed children', async () => {
      /** Arrange */
      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider>
              <TestChild />
            </AlertProvider>
          </ThemeConfig>
        </Provider>
      );

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.queryByText('test child')).toBeNull();
      });
    });

    it('Should render no children if non-passed', async () => {
      /** Arrange */
      /** Act */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider />
          </ThemeConfig>
        </Provider>
      );

      /** Assert */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.queryByText('test child')).toBeNull();
      });
    });
  });

  describe('Message Box', () => {
    it('Should render the passed severity type', async () => {
      /** Arrange - create "success" notification */
      store.dispatch(
        addNotification({
          message: testMessage,
          variant: 'success',
        })
      );

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider>
              <TestChild />
            </AlertProvider>
          </ThemeConfig>
        </Provider>
      );

      /** Assert - check that alert has been rendered and is success colour */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByRole('alert')).toBeInTheDocument();
      });
      expect(screen.getByRole('alert')).toHaveStyle(
        `background-color: ${palette.success.main}`
      );
      expect(screen.getByText(testMessage)).toBeInTheDocument();

      /** Act - close the alert notification */
      fireEvent(
        screen.getByTestId('CloseIcon'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );

      /** Assert - check that alert has closed */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.queryByRole('alert')).toBeNull();
      });

      /** Act - create "error" notification */
      act(() => {
        store.dispatch(
          addNotification({
            message: testMessage,
            variant: 'error',
          })
        );
      });

      /** Assert - check that alert has been rendered and is error colour */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByRole('alert')).toBeInTheDocument();
      });
      expect(screen.getByRole('alert')).toHaveStyle(
        `background-color: ${palette.error.main}`
      );
    });

    it('Should close when close button is clicked', async () => {
      /** Arrange - create "success" notification */
      store.dispatch(
        addNotification({
          message: testMessage,
          variant: 'success',
        })
      );

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider>
              <TestChild />
            </AlertProvider>
          </ThemeConfig>
        </Provider>
      );

      /** Assert - check that the alert has been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.getByRole('alert')).toBeInTheDocument();
      });
      expect(screen.getByText(testMessage)).toBeInTheDocument();

      /** Act - close the alert notification */
      fireEvent(
        screen.getByTestId('CloseIcon'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );

      /** Assert - check that notification has been removed */
      await waitFor(() => {
        /** After all state updates have completed */
        expect(screen.queryByRole('alert')).toBeNull();
      });
    });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      addNotification({
        message: testMessage,
        variant: 'success',
      })
    );
    store.dispatch(
      addNotification({
        message: testMessage,
        variant: 'error',
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <AlertProvider>
              <TestChild />
            </AlertProvider>
          </ThemeConfig>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
