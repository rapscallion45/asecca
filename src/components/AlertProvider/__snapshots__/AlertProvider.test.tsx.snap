// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Provider Renders correctly 1`] = `
<div>
  Test child
</div>
`;
